File: storage/ndb/clusterj/clusterj-jdbc/src/test/java/jdbctest/BatchTest.java
/*
   Copyright 2011, Oracle and/or its affiliates. All rights reserved.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; version 2 of the License.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

package jdbctest;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import testsuite.clusterj.AbstractClusterJModelTest;

public class BatchTest extends AbstractClusterJModelTest {

    private static final int NUMBER_OF_INSTANCES = 10;

    private static final Set<Integer> expecteds = new HashSet<Integer>();
    static {
        for (int i = 0; i < NUMBER_OF_INSTANCES;++i) {
            expecteds.add(i);
        }
    }
    @Override
    public boolean getDebug() {
        return false;
    }

    @Override
    public void localSetUp() {
        super.localSetUp();
    }

    public void testInsertBatch() {
        try {
            connection.setAutoCommit(false);
            deleteAll(connection);
            int[] counts = insertBatch(connection);
            if (getDebug()) System.out.println(Arrays.toString(counts));
            for (int i = 0; i < NUMBER_OF_INSTANCES; ++i) {
                int count = counts[i];
                errorIfNotEqual("test executeBatch failure for " + i, 1, count);
            }
            connection.commit();
            selectAndVerifyAll(connection);
            closeConnection();
        } catch (SQLException e) {
            error("insert id, name, age, magic into t_basic values(?, ?, ?, ?) threw " + e.getMessage());
            if (getDebug()) e.printStackTrace();
            throw new RuntimeException("insert id, name, age, magic into t_basic values(?, ?, ?, ?)", e);
        } catch (RuntimeException e) {
            if (getDebug()) e.printStackTrace();
        }
        failOnError();
    }

    public void testInsertBatchAutocommit() {
        try {
            connection.setAutoCommit(false);
            deleteAll(connection);
            connection.setAutoCommit(true);
            insertBatch(connection);
            closeConnection();
        } catch (SQLException e) {
            error("insert id, name, age, magic into t_basic values(?, ?, ?, ?) threw " + e.getMessage());
            if (getDebug()) e.printStackTrace();
            throw new RuntimeException("insert id, name, age, magic into t_basic values(?, ?, ?, ?)", e);
        } catch (RuntimeException e) {
            if (getDebug()) e.printStackTrace();
        }
        failOnError();
    }

    private int deleteAll(Connection connection) throws SQLException {
        Statement deleteStatement = connection.createStatement();
        int result = deleteStatement.executeUpdate("delete from t_basic");
        deleteStatement.close();
        connection.commit();
        return result;
    }

    private int[] insertBatch(Connection connection) {
        PreparedStatement statement;
        int[] counts =  null;
        try {
            statement = connection.prepareStatement(
                    "insert into t_basic (id, name, age, magic) values(?, ?, ?, ?)");
            for (int i = 0; i < NUMBER_OF_INSTANCES; ++i) {
                statement.setInt(1, i);
                statement.setString(2, "Employee " + i);
                statement.setInt(3, i);
                statement.setInt(4, i);
                statement.addBatch();
            }
            counts = statement.executeBatch();
            statement.close();
            return counts;
        } catch (SQLException e) {
            throw new RuntimeException("insertBatch.executeBatch threw " + e.getMessage(), e);
        }
        
    }

    private void selectAndVerifyAll(Connection connection) throws SQLException {
        PreparedStatement selectStatement = connection.prepareStatement(
                "select id, name, age, magic from t_basic");
        ResultSet rs = selectStatement.executeQuery();
        Set<Integer> actuals = new HashSet<Integer>();
        while (rs.next()) {
            int id = rs.getInt(1);
            verifyEmployee(rs, id);
            actuals.add(id);
        }
        errorIfNotEqual("Wrong number of instances in database.", expecteds, actuals);
    }

    private void verifyEmployee(ResultSet rs, int id) {
        try {
            String name = rs.getString(2);
            errorIfNotEqual("Verify name id: " + id, "Employee " + id, name);
            int age = rs.getInt(3);
            errorIfNotEqual("Verify age id: " + id, id, age);
            int magic = rs.getInt(4);
            errorIfNotEqual("Verify magic id: " + id, id, magic);
        } catch (SQLException e) {
            if (getDebug()) e.printStackTrace();
        }
    }

}

File: test/t_basic/BatchTest.js
/*
 Copyright (c) 2012, Oracle and/or its affiliates. All rights
 reserved.
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; version 2 of
 the License.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 02110-1301  USA
 */

/** Dummy test */
t0 = new harness.SerialTest('dummy');
t0.run = function() {
  this.pass();
};

udebug      = unified_debug.getLogger("BatchTest.js");

function createBatch(session, start_value, number, testCase) {
  udebug.log("createBatch");
  var batch = session.createBatch();
  var i;
  var object;
  for (i = start_value; i < start_value + number; ++i) {
    object = new global.t_basic(i, 'Employee ' + i, i, i);
    batch.persist(object, function(err, callback_session, callback_i, callback_testCase) {
      if (err && !testCase.expectError) {
        callback_testCase.appendErrorMessage('Error inserting ' + callback_i + JSON.stringify(err));
      }
    }, session, i, testCase);
  }
  return batch;
};

function onExecuteBatch(err, session, start_value, number, testCase) {
  udebug.log("onExecuteBatch");
  if (err) {
    testCase.fail(err);
    return;
  }
  var tx = session.currentTransaction();
  if (tx.isActive()) {
    tx.commit(verifyInsert, session, start_value, number, testCase);
  }
};

function verifyInsert(err, session, start_value, number, testCase) {
  udebug.log('verifyInsert');
  if (err) {
    testCase.fail(err);
    return;
  }
  // after the batch insert, verify that the insert occurred
  var j;
  var completed = 0;
  for (j = start_value; j < start_value + number; ++j) {
    session.find(global.t_basic, j, function(err, found, testCase, callback_j, session) {
      var tx = session.currentTransaction();
      if (err && !testCase.expectError) {
        testCase.appendErrorMessage(
            'BatchTest.verifyInsert find callback for ', j,' failed: ', JSON.stringify(err));
        return;
      }
      found_j = 'undefined';
      if (found !== null) {
        found_j = found.id;
        if (found_j !== callback_j) {
          testCase.errorIfNotEqual(
              'BatchTest.verifyInsert find callback failed', found_j, found.id);
        }
      }
      udebug.log('BatchTest.verifyInsert find callback for ', found_j);
      if (++completed === number) {
        // end of test case; all callbacks completed
        if (tx.isActive()) {
          tx.commit(function(err, session, testCase) {
            if (err) {
              testCase.fail(err);
              return;
            }
            testCase.failOnError();
          }, session, testCase);
        } else {
          testCase.failOnError();
        }
      }
    }, testCase, j, session);
  }
};

/***** Insert Autocommit ***/
var t1 = new harness.SerialTest('testBatchInsertAutocommit');
t1.run = function() {
  var testCase = this;
  this.number_to_insert = 10;
  this.start_value = 10000;
  fail_openSession(testCase, function(session) {
    var batch = createBatch(session, testCase.start_value, testCase.number_to_insert, testCase);
    batch.execute(verifyInsert, session, testCase.start_value, testCase.number_to_insert, testCase);
  });
};

/***** Insert begin commit ***/
var t2 = new harness.SerialTest('testBatchInsertBeginCommit');
t2.run = function() {
  var testCase = this;
  this.number_to_insert = 10;
  this.start_value = 20000;
  fail_openSession(testCase, function(session) {
    var batch = createBatch(session, testCase.start_value, testCase.number_to_insert, testCase);
    session.currentTransaction().begin();
    batch.execute(onExecuteBatch, session, testCase.start_value, testCase.number_to_insert, testCase);
  });
};

/***** getSession ***/
var t3 = new harness.ConcurrentTest('getSession');
t3.run = function() {
  var testCase = this;
  fail_openSession(testCase, function(session) {
    var batch = session.createBatch();
    if (batch.getSession() !== session) {
      testCase.fail(new Error('getSession failed to return session'));
    } else {
      testCase.pass();
    }
  });
};

/***** clear ***/
var t4 = new harness.ConcurrentTest('clear');
t4.expectError = true;
t4.run = function() {
  var testCase = this;
  this.number_to_insert = 10;
  // these values conflict with t1 but these values are cleared, not committed
  this.start_value = 10000;
  fail_openSession(testCase, function(session) {
    var batch = createBatch(session, testCase.start_value, testCase.number_to_insert, testCase);
    batch.clear();
    batch.execute(function(err) {
      if (err) {
        testCase.appendError(err);
      }
    });
    testCase.failOnError();
  });
};


/*************** EXPORT THE TOP-LEVEL GROUP ********/
// module.exports.tests = [t1, t2, t3, t4];
module.exports.tests = [t1];
